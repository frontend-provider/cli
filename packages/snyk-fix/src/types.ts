import { DepGraphData } from '@snyk/dep-graph';
import { CustomError } from './lib/errors/custom-error';

/* Scan Result
 * this data is returned by the CLI plugins to identify
 * what should be scanned for issues
 */
export interface GitTarget {
  remoteUrl: string;
  branch: string;
}
export interface ContainerTarget {
  image: string;
}

export interface ScanResult {
  readonly identity: Identity;
  readonly facts: Facts[];
  readonly name?: string;
  readonly policy?: string;
  readonly target?: GitTarget | ContainerTarget;
}

export interface Identity {
  type: string;
  targetFile?: string;
  // options used to scan should be here
  args?: { [key: string]: string };
}

export interface Facts {
  type: string;
  data: any;
}

/* Test Result
 * this data is returned on a `snyk test` for supported project types
 * after the relevant plugin extracts dependencies
 */
export interface TestResult {
  readonly issues: Issue[];
  readonly issuesData: IssuesData;
  readonly depGraphData: DepGraphData;
  readonly remediation?: RemediationChanges; // TODO: not yet in the CLI TestResults type
}

export interface Issue {
  pkgName: string;
  pkgVersion?: string;
  issueId: string;
  fixInfo: FixInfo;
}

interface UpgradePath {
  path: UpgradePathItem[];
}

export interface FixInfo {
  upgradePaths?: UpgradePath[];
  isPatchable?: boolean;
  nearestFixedInVersion?: string;
}

interface UpgradePathItem {
  name: string;
  version: string;
  newVersion?: string;
  isDropped?: boolean;
}
export interface IssuesData {
  [issueId: string]: {
    id: string;
    severity: SEVERITY;
    title: string;
  };
}

/* Remediation Data
 * this data is returned on a `snyk test` for supported project types
 */
export interface Upgrade {
  upgradeTo: string; // name@version
}

export interface UpgradeVulns extends Upgrade {
  vulns: string[];
}
export interface UpgradeRemediation extends UpgradeVulns {
  upgrades: string[];
}

export interface PatchRemediation {
  paths: PatchObject[];
}

export interface DependencyUpdates {
  [from: string]: UpgradeRemediation;
}

export interface PinRemediation extends UpgradeVulns {
  isTransitive: boolean;
}

export interface DependencyPins {
  [name: string]: PinRemediation;
}

// Remediation changes to be applied to the project,
// including information on all and unresolved issues.
export interface RemediationChanges {
  unresolved: IssueData[];
  upgrade: DependencyUpdates;
  patch: {
    [name: string]: PatchRemediation;
  };
  ignore: unknown;
  pin: DependencyPins;
}

export interface IssueData {
  id: string;
  packageName: string;
  version: string;
  moduleName?: string;
  below: string; // Vulnerable below version
  semver: {
    vulnerable: string | string[];
    vulnerableHashes?: string[];
    vulnerableByDistro?: {
      [distroNameAndVersion: string]: string[];
    };
  };
  patches: Patch[];
  isNew: boolean;
  description: string;
  title: string;
  severity: SEVERITY;
  fixedIn: string[];
  legalInstructions?: string;
  reachability?: REACHABILITY;
}

interface Patch {
  version: string;
  id: string;
  urls: string[];
  modificationTime: string;
}

export enum REACHABILITY {
  FUNCTION = 'function',
  PACKAGE = 'package',
  NOT_REACHABLE = 'not-reachable',
  NO_INFO = 'no-info',
}

export interface PatchObject {
  [name: string]: {
    patched: string;
  };
}

export enum SEVERITY {
  LOW = 'low',
  MEDIUM = 'medium',
  HIGH = 'high',
  CRITICAL = 'critical',
}

/* End Remediation Data
 */

/* Snyk fix types
 * Types for concepts introduced as part of this lib
 */

export type SupportedScanTypes = 'pip';

export interface Workspace {
  path: string;
  readFile: (path: string) => Promise<string>;
  writeFile: (path: string, content: string) => Promise<void>;
}
export interface EntityToFix {
  readonly workspace: Workspace;
  readonly scanResult: ScanResult;
  readonly testResult: TestResult;
  readonly options: CliTestOptions;
}

// Partial CLI test options interface
// defining only what is used by @snyk/fix
// add more as needed
export interface PythonTestOptions {
  command?: string; // python interpreter to use for python tests
  dev?: boolean;
}
export type CliTestOptions = PythonTestOptions;
export interface WithError<Original> {
  original: Original;
  error: CustomError;
  tip?: string;
}

export interface WithFixChangesApplied<Original> {
  original: Original;
  changes: FixChangesSummary[];
}

export interface WithUserMessage<Original> {
  original: Original;
  userMessage: string;
}

export type FixChangesSummary = FixChangesSuccess | FixChangesError;

interface FixChangesSuccess {
  success: true;
  userMessage: string;
  issueIds: string[];
  from?: string;
  to?: string;
}

interface FixChangesError {
  success: false;
  userMessage: string;
  reason: string;
  tip?: string;
  issueIds: string[];
  from?: string;
  to?: string;
}

export interface ErrorsByEcoSystem {
  [ecosystem: string]: { originals: EntityToFix[]; userMessage: string };
}
export interface FixOptions {
  dryRun?: boolean;
  quiet?: boolean;
  stripAnsi?: boolean;
}

export interface FixedMeta {
  fixed: number;
  failed: number;
  fixableIssues: number;
  fixedIssues: number;
  totalIssues: number;
}
