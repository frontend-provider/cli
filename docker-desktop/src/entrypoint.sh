#!/usr/bin/env bash
set -euo pipefail

script_dir="$(dirname "$0")"
"${script_dir}/node-release/bin/node" "${script_dir}/bin/snyk" "$@"
