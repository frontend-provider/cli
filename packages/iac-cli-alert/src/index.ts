import {
  IncomingWebhook,
  IncomingWebhookDefaultArguments,
} from '@slack/webhook';

if (!process.env.IAC_SMOKE_TESTS_SLACK_WEBHOOK_URL) {
  console.error(
    'Missing the IAC_SMOKE_TESTS_SLACK_WEBHOOK_URL environment variable',
  );
  process.exit(1);
}

if (!process.env.IAC_SMOKE_TESTS_OS) {
  console.error('Missing the IAC_SMOKE_TESTS_OS environment variable');
  process.exit(1);
}

const IAC_SMOKE_TESTS_SLACK_WEBHOOK_URL =
  process.env.IAC_SMOKE_TESTS_SLACK_WEBHOOK_URL;
const IAC_SMOKE_TESTS_OS = process.env.IAC_SMOKE_TESTS_OS;

const slackWebhook = new IncomingWebhook(IAC_SMOKE_TESTS_SLACK_WEBHOOK_URL);

async function sendSlackAlert() {
  console.log('IaC smoke tests failed. Sending Slack alert...');
  const args: IncomingWebhookDefaultArguments = {
    username: 'IaC CLI Alerts',
    text: `An Infrastructure as Code Smoke Tests job failed. \n Operating system - ${IAC_SMOKE_TESTS_OS} \n<https://github.com/snyk/cli/actions?query=workflow%3A+%22Infrastructure+as+Code+Smoke+Tests%22|Infrastructure as Code Smoke Tests Results>`,
    icon_emoji: 'snyk-iac',
  };
  await slackWebhook.send(args);
  console.log('Slack alert sent.');
}

async function run() {
  try {
    await sendSlackAlert();
  } catch (error) {
    console.error(error);
    process.exit(1);
  }
}

process.on('uncaughtException', (err) => {
  console.error(err);
  process.exit(1);
});

// Exec
run();
