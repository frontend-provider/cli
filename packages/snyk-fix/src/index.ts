import * as debugLib from 'debug';
import { loadPlugin } from './plugins/load-plugin';
import { FixHandlerResult } from './plugins/types';

import { EntityToFix } from './types';

const debug = debugLib('snyk-fix:main');

export async function fix(
  entities: EntityToFix[],
): Promise<FixHandlerResult[]> {
  debug(`Requested to fix ${entities.length} projects.`);
  const entitiesPerType = groupEntitiesPerType(entities);
  const allResults: FixHandlerResult[] = [];
  // TODO: pMap this?
  for (const type of Object.keys(entitiesPerType)) {
    const handler = loadPlugin(type);
    const results = await handler(entitiesPerType[type]);
    allResults.push(results);
  }

  return allResults;
}

function groupEntitiesPerType(
  entities: EntityToFix[],
): {
  [type: string]: EntityToFix[];
} {
  const entitiesPerType: {
    [type: string]: EntityToFix[];
  } = {};
  for (const entity of entities) {
    const type = entity.scanResult.identity.type;
    if (entitiesPerType[type]) {
      entitiesPerType[type].push(entity);
      continue;
    }
    entitiesPerType[type] = [entity];
  }
  return entitiesPerType;
}
