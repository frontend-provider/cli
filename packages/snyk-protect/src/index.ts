#!/usr/bin/env node

import protect from './lib';

async function main() {
  const projectPath = process.cwd();
  await protect(projectPath);
}

if (require.main === module) {
  main();
}
