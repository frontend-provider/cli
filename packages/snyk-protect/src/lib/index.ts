import * as fs from 'fs';
import * as path from 'path';
import { extractPatchMetadata } from './snyk-file';
import { applyPatchToFile } from './patch';
import { findPhysicalModules } from './explore-node-modules';
import {
  VulnIdAndPackageName,
  VulnPatches,
  PatchedModule,
  ProtectResultType,
} from './types';
import { getAllPatches } from './fetch-patches';
import { sendAnalytics } from './analytics';

async function protect(projectFolderPath: string) {
  const snykFilePath = path.resolve(projectFolderPath, '.snyk');

  if (!fs.existsSync(snykFilePath)) {
    console.log('No .snyk file found');
    sendAnalytics({
      type: ProtectResultType.NO_SNYK_FILE,
    });
    return;
  }

  const snykFileContents = fs.readFileSync(snykFilePath, 'utf8');
  const snykFilePatchMetadata = extractPatchMetadata(snykFileContents);
  const vulnIdAndPackageNames: VulnIdAndPackageName[] = snykFilePatchMetadata;
  const targetPackageNames = [
    ...new Set(snykFilePatchMetadata.map((vpn) => vpn.packageName)), // get a list of unique package names by converting to Set and then back to array
  ];

  // find instances of the target packages by spelunking through the node_modules looking for modules with a target packageName
  const foundPhysicalPackages = findPhysicalModules(
    projectFolderPath,
    targetPackageNames,
  );

  // Map of package name to versions (for the target package names).
  // For each package name, we might have found multiple versions and we'll need to fetch patches for each version.
  // We will use this to lookup the versions of packages to get patches for.
  const packageNameToVersionsMap = new Map<string, string[]>();
  foundPhysicalPackages.forEach((p) => {
    if (packageNameToVersionsMap.has(p.packageName)) {
      const versions = packageNameToVersionsMap.get(p.packageName);
      if (!versions?.includes(p.packageVersion)) {
        versions?.push(p.packageVersion);
      }
    } else {
      packageNameToVersionsMap.set(p.packageName, [p.packageVersion]);
    }
  });

  const packageAtVersionsToPatches: Map<
    string,
    VulnPatches[]
  > = await getAllPatches(vulnIdAndPackageNames, packageNameToVersionsMap);

  if (packageAtVersionsToPatches.size === 0) {
    console.log('Nothing to patch');
    sendAnalytics({
      type: ProtectResultType.NOTHING_TO_PATCH,
    });
    return;
  }

  const patchedModules: PatchedModule[] = [];
  foundPhysicalPackages.forEach((fpp) => {
    const packageNameAtVersion = `${fpp.packageName}@${fpp.packageVersion}`;
    const vuldIdAndPatches = packageAtVersionsToPatches.get(
      packageNameAtVersion,
    );
    vuldIdAndPatches?.forEach((vp) => {
      vp.patches.forEach((patchDiffs) => {
        patchDiffs.patchDiffs.forEach((diff) => {
          applyPatchToFile(diff, fpp.path);
        });
      });
      patchedModules.push({
        vulnId: vp.vulnId,
        packageName: fpp.packageName,
        packageVersion: fpp.packageVersion,
      });
    });
  });

  console.log('Successfully applied Snyk patches');

  sendAnalytics({
    type: ProtectResultType.APPLIED_PATCHES,
    patchedModules,
  });
}

export default protect;
