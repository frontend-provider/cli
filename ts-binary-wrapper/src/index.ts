#!/usr/bin/env node

import { argv } from 'process';
import * as common from './common';

const config = common.getCurrentConfiguration();
const executable = config.getLocalLocation();
const cliArguments = common.getCliArguments(argv);
const exitCode = common.runWrapper(executable, cliArguments);
process.exit(exitCode);
