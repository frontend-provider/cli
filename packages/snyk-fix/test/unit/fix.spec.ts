import * as pipenvPipfileFix from '@snyk/fix-pipenv-pipfile';

import * as snykFix from '../../src';
import { generateEntityToFix } from '../helpers/generate-entity-to-fix';

jest.mock('@snyk/fix-pipenv-pipfile');

describe('Snyk fix', () => {
  beforeAll(() => {
    jest.spyOn(pipenvPipfileFix, 'isPipenvSupportedVersion').mockReturnValue({
      supported: true,
      versions: ['123.123.123'],
    });
    jest.spyOn(pipenvPipfileFix, 'isPipenvInstalled').mockResolvedValue({
      version: '123.123.123',
    });
    jest.spyOn(pipenvPipfileFix, 'pipenvInstall').mockResolvedValue({
      exitCode: 0,
      stdout: '',
      stderr: '',
      command: 'pipenv install',
      duration: 123,
    });
  });

  it('Snyk fix returns results for supported type', async () => {
    // Arrange
    const projectTestResult = generateEntityToFix(
      'pip',
      'requirements.txt',
      'django===1.6.1',
    );
    const writeFileSpy = jest.spyOn(projectTestResult.workspace, 'writeFile');

    // Act
    const res = await snykFix.fix([projectTestResult], {
      quiet: true,
      stripAnsi: true,
    });

    // Assert
    expect(writeFileSpy).toHaveBeenCalledTimes(1);
    expect(res.exceptions).toMatchSnapshot();
    expect(res.results).toMatchSnapshot();
  });

  it('Snyk fix returns results for supported type in dryRun mode (no write)', async () => {
    // Arrange
    const projectTestResult = generateEntityToFix(
      'pip',
      'requirements.txt',
      'django===1.6.1',
    );
    const writeFileSpy = jest.spyOn(projectTestResult.workspace, 'writeFile');
    // Act
    await snykFix.fix([projectTestResult], {
      quiet: true,
      dryRun: true,
    });

    // Assert
    expect(writeFileSpy).not.toHaveBeenCalled();
  });

  it('Snyk fix returns results for supported & unsupported type', async () => {
    // Arrange
    const projectTestResult = generateEntityToFix(
      'pip',
      'requirements.txt',
      'django===1.6.1',
    );
    const pipfileProjectTestResult = generateEntityToFix('pip', 'Pipfile', '');

    // Act
    const res = await snykFix.fix(
      [projectTestResult, pipfileProjectTestResult],
      { quiet: true, stripAnsi: true, dryRun: true },
    );

    // Assert
    expect(res).toMatchSnapshot();
  });

  it('Snyk fix returns results as expected', async () => {
    // Arrange
    const txtProdProjectTestResult = generateEntityToFix(
      'pip',
      'prod.txt',
      'django===1.6.1',
    );
    const txtDevProjectTestResult = generateEntityToFix(
      'pip',
      'dev.txt',
      'django===1.6.1',
    );
    const pipfileProjectTestResult = generateEntityToFix('pip', 'Pipfile', '');

    // Act
    const res = await snykFix.fix(
      [
        txtDevProjectTestResult,
        txtProdProjectTestResult,
        pipfileProjectTestResult,
      ],
      { quiet: true, stripAnsi: true, dryRun: true },
    );

    // Assert
    expect(res.exceptions).toEqual({});
    expect(Object.keys(res.results)).toHaveLength(1);
    expect(Object.keys(res.results)[0]).toEqual('python');
    // skipped unsupported
    expect(res.results.python.skipped).toHaveLength(0);

    // first *.txt throws because of the mock above
    expect(res.results.python.failed).toHaveLength(0);
    expect(res.results.python.succeeded).toHaveLength(3);
    expect(
      res.results.python.succeeded[0].original.scanResult.identity.targetFile,
    ).toEqual('Pipfile');
    expect(
      res.results.python.succeeded[1].original.scanResult.identity.targetFile,
    ).toEqual('dev.txt');
    expect(
      res.results.python.succeeded[2].original.scanResult.identity.targetFile,
    ).toEqual('prod.txt');
  });
  it('Snyk fix returns results as expected when 1 fails to fix', async () => {
    // Arrange
    const txtProdProjectTestResult = generateEntityToFix(
      'pip',
      'prod.txt',
      'django===1.6.1',
    );
    const txtDevProjectTestResult = generateEntityToFix(
      'pip',
      'dev.txt',
      'django===1.6.1',
    );
    jest
      .spyOn(txtDevProjectTestResult.workspace, 'readFile')
      .mockImplementation(() => {
        throw new Error('Test Error: Invalid encoding');
      });
    const pipfileProjectTestResult = generateEntityToFix('pip', 'Pipfile', '');

    // Act
    const res = await snykFix.fix(
      [
        txtDevProjectTestResult,
        txtProdProjectTestResult,
        pipfileProjectTestResult,
      ],
      { quiet: true, stripAnsi: true, dryRun: true },
    );

    // Assert
    expect(res.exceptions).toEqual({});
    expect(Object.keys(res.results)).toHaveLength(1);
    expect(Object.keys(res.results)[0]).toEqual('python');

    expect(res.results.python.skipped).toHaveLength(0);

    expect(res.results.python.failed[0]).toEqual({
      error: new Error('Test Error: Invalid encoding'),
      original: txtDevProjectTestResult,
    });

    expect(res.results.python.failed).toHaveLength(1);
    expect(res.results.python.succeeded).toHaveLength(2);

    expect(
      res.results.python.succeeded[0].original.scanResult.identity.targetFile,
    ).toEqual('Pipfile');
    expect(
      res.results.python.succeeded[1].original.scanResult.identity.targetFile,
    ).toEqual('prod.txt');
  });

  it('Snyk fix returns results as expected when remediation data is empty', async () => {
    // Arrange
    const txtProdProjectTestResult = generateEntityToFix(
      'pip',
      'prod.txt',
      'django===1.6.1',
    );
    // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
    // @ts-ignore: for test purpose only
    delete txtProdProjectTestResult.testResult.remediation;

    // Act
    const res = await snykFix.fix([txtProdProjectTestResult], {
      quiet: true,
      stripAnsi: true,
    });
    // Assert
    expect(res.exceptions).toEqual({});
    expect(Object.keys(res.results)).toHaveLength(1);
    expect(Object.keys(res.results)[0]).toEqual('python');

    // first *.txt throws because remediation is empty
    expect(res.results.python.failed).toHaveLength(0);
    expect(res.results.python.skipped).toHaveLength(1);
    expect(
      res.results.python.skipped[0].original.scanResult.identity.targetFile,
    ).toEqual('prod.txt');
    expect(res.results.python.skipped[0].userMessage).toEqual(
      'No remediation data available',
    );
  });
});

describe('groupEntitiesPerScanType', () => {
  it('It correctly groups related entities per handler type (pip)', () => {
    // Arrange
    const txtProdProjectTestResult = generateEntityToFix(
      'pip',
      'prod.txt',
      'django===1.6.1',
    );
    const txtDevProjectTestResult = generateEntityToFix(
      'pip',
      'dev.txt',
      'django===1.6.1',
    );
    const pipfileProjectTestResult = generateEntityToFix(
      'pip',
      'Pipfile',
      'django===1.6.1',
    );

    // Act
    const res = snykFix.groupEntitiesPerScanType([
      txtProdProjectTestResult,
      txtDevProjectTestResult,
      pipfileProjectTestResult,
    ]);

    // Assert
    expect(Object.keys(res)[0]).toEqual('pip');
    expect(Object.keys(res)[0]).toHaveLength(3);
  });
  it('It correctly groups related entities per handler type (mixed)', () => {
    // Arrange
    const txtProdProjectTestResult = generateEntityToFix(
      'pip',
      'prod.txt',
      'django===1.6.1',
    );
    const txtDevProjectTestResult = generateEntityToFix(
      'pip',
      'dev.txt',
      'django===1.6.1',
    );
    const npmProjectTestResult = generateEntityToFix(
      'npm',
      'package.json',
      'django===1.6.1',
    );

    const poetryProjectTestResult = generateEntityToFix(
      'poetry',
      'poetry.lock',
      'django===1.6.1',
    );

    // Act
    const res = snykFix.groupEntitiesPerScanType([
      txtProdProjectTestResult,
      txtDevProjectTestResult,
      npmProjectTestResult,
      poetryProjectTestResult,
    ]);

    // Assert
    expect(Object.keys(res).sort()).toEqual(['npm', 'pip', 'poetry']);
    expect(res.npm).toHaveLength(1);
    expect(res.pip).toHaveLength(2);
    expect(res.poetry).toHaveLength(1);
  });

  it('It correctly groups related entities per handler type with missing type', () => {
    // Arrange
    const txtProdProjectTestResult = generateEntityToFix(
      'pip',
      'prod.txt',
      'django===1.6.1',
    );
    const txtDevProjectTestResult = generateEntityToFix(
      'pip',
      'dev.txt',
      'django===1.6.1',
    );
    const missingProjectTestResult = generateEntityToFix(
      'npm',
      'package.json',
      'django===1.6.1',
    );
    // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
    // @ts-ignore: for test purpose only
    delete missingProjectTestResult.scanResult.identity.type;

    // Act
    const res = snykFix.groupEntitiesPerScanType([
      txtProdProjectTestResult,
      txtDevProjectTestResult,
      missingProjectTestResult,
    ]);

    // Assert
    expect(Object.keys(res).sort()).toEqual(['missing-type', 'pip']);
    expect(res['missing-type']).toHaveLength(1);
    expect(res.pip).toHaveLength(2);
  });
});

describe('Error handling', () => {
  it('Snyk fix returns error when called with unsupported type', async () => {
    // Arrange
    const projectTestResult = generateEntityToFix(
      'npm',
      'package.json',
      JSON.stringify({}),
    );
    // Act
    const res = await snykFix.fix([projectTestResult], {
      quiet: true,
      stripAnsi: true,
    });
    // Assert
    expect(res).toMatchSnapshot();
  });

  it('Snyk fix returns error when manifest can not be parsed', async () => {
    // Arrange
    const projectTestResult = generateEntityToFix(
      'pip',
      'requirements.txt',
      'django@===1.6.1', // invalid requirements
    );
    // Act
    const res = await snykFix.fix([projectTestResult], {
      quiet: true,
      stripAnsi: true,
    });
    // Assert
    expect(res).toMatchSnapshot();
  });
});
