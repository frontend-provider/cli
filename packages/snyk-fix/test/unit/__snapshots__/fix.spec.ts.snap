// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Error handling Snyk fix returns error when called with unsupported type 1`] = `
Object {
  "exceptions": Object {
    "npm": Object {
      "originals": Array [
        Object {
          "options": Object {
            "command": "python3",
          },
          "scanResult": Object {
            "facts": Array [
              Object {
                "data": "not-implemented",
                "type": "not-implemented",
              },
            ],
            "identity": Object {
              "targetFile": "package.json",
              "type": "npm",
            },
          },
          "testResult": Object {
            "depGraphData": "",
            "issues": Array [
              Object {
                "fixInfo": Object {},
                "issueId": "VULN_ID_1",
                "pkgName": "package@version",
              },
            ],
            "issuesData": Object {
              "vuln-id": Object {
                "id": "VULN_ID_1",
                "severity": "high",
                "title": "Fake vuln",
              },
            },
            "remediation": Object {
              "ignore": Object {},
              "patch": Object {},
              "pin": Object {
                "django@1.6.1": Object {
                  "isTransitive": false,
                  "upgradeTo": "django@2.0.1",
                  "vulns": Array [
                    "vuln-id",
                  ],
                },
              },
              "unresolved": Array [],
              "upgrade": Object {},
            },
          },
          "workspace": Object {
            "path": ".",
            "readFile": [Function],
            "writeFile": [Function],
          },
        },
      ],
      "userMessage": "npm is not supported.",
    },
  },
  "fixSummary": "
 ✖ No successful fixes


Unresolved items:

  package.json
  ✖ npm is not supported.

Summary:

  1 items were not fixed

  1 issues: 1 High
  1 issues are fixable
  

Tip: Re-run in debug mode to see more information: DEBUG=*snyk* <COMMAND>. If the issue persists contact support@snyk.io",
  "meta": Object {
    "failed": 1,
    "fixableIssues": 1,
    "fixed": 0,
    "fixedIssues": 0,
    "totalIssues": 1,
  },
  "results": Object {},
}
`;

exports[`Error handling Snyk fix returns error when manifest can not be parsed 1`] = `
Object {
  "exceptions": Object {},
  "fixSummary": "
Successful fixes:

  requirements.txt
  ✔ Pinned django from 1.6.1 to 2.0.1

Summary:

  1 items were successfully fixed

  1 issues: 1 High
  1 issues are fixable
  1 issues were successfully fixed
",
  "meta": Object {
    "failed": 0,
    "fixableIssues": 1,
    "fixed": 1,
    "fixedIssues": 1,
    "totalIssues": 1,
  },
  "results": Object {
    "python": Object {
      "failed": Array [],
      "skipped": Array [],
      "succeeded": Array [
        Object {
          "changes": Array [
            Object {
              "from": "django@1.6.1",
              "issueIds": Array [
                "vuln-id",
              ],
              "success": true,
              "to": "django@2.0.1",
              "userMessage": "Pinned django from 1.6.1 to 2.0.1",
            },
          ],
          "original": Object {
            "options": Object {
              "command": "python3",
            },
            "scanResult": Object {
              "facts": Array [
                Object {
                  "data": "not-implemented",
                  "type": "not-implemented",
                },
              ],
              "identity": Object {
                "targetFile": "requirements.txt",
                "type": "pip",
              },
            },
            "testResult": Object {
              "depGraphData": "",
              "issues": Array [
                Object {
                  "fixInfo": Object {},
                  "issueId": "VULN_ID_1",
                  "pkgName": "package@version",
                },
              ],
              "issuesData": Object {
                "vuln-id": Object {
                  "id": "VULN_ID_1",
                  "severity": "high",
                  "title": "Fake vuln",
                },
              },
              "remediation": Object {
                "ignore": Object {},
                "patch": Object {},
                "pin": Object {
                  "django@1.6.1": Object {
                    "isTransitive": false,
                    "upgradeTo": "django@2.0.1",
                    "vulns": Array [
                      "vuln-id",
                    ],
                  },
                },
                "unresolved": Array [],
                "upgrade": Object {},
              },
            },
            "workspace": Object {
              "path": ".",
              "readFile": [Function],
              "writeFile": [Function],
            },
          },
        },
      ],
    },
  },
}
`;

exports[`Snyk fix Snyk fix returns results for supported & unsupported type 1`] = `
Object {
  "exceptions": Object {},
  "fixSummary": "
Successful fixes:

  Pipfile
  ✔ Upgraded django from 1.6.1 to 2.0.1

  requirements.txt
  ✔ Upgraded django from 1.6.1 to 2.0.1

Summary:

  Command run in dry run mode. Fixes are not applied.
  2 items were successfully fixed

  2 issues: 2 High
  2 issues are fixable
  2 issues were successfully fixed
",
  "meta": Object {
    "failed": 0,
    "fixableIssues": 2,
    "fixed": 2,
    "fixedIssues": 2,
    "totalIssues": 2,
  },
  "results": Object {
    "python": Object {
      "failed": Array [],
      "skipped": Array [],
      "succeeded": Array [
        Object {
          "changes": Array [
            Object {
              "from": "django@1.6.1",
              "issueIds": Array [
                "vuln-id",
              ],
              "success": true,
              "to": "django@2.0.1",
              "userMessage": "Upgraded django from 1.6.1 to 2.0.1",
            },
          ],
          "original": Object {
            "options": Object {
              "command": "python3",
            },
            "scanResult": Object {
              "facts": Array [
                Object {
                  "data": "not-implemented",
                  "type": "not-implemented",
                },
              ],
              "identity": Object {
                "targetFile": "Pipfile",
                "type": "pip",
              },
            },
            "testResult": Object {
              "depGraphData": "",
              "issues": Array [
                Object {
                  "fixInfo": Object {},
                  "issueId": "VULN_ID_1",
                  "pkgName": "package@version",
                },
              ],
              "issuesData": Object {
                "vuln-id": Object {
                  "id": "VULN_ID_1",
                  "severity": "high",
                  "title": "Fake vuln",
                },
              },
              "remediation": Object {
                "ignore": Object {},
                "patch": Object {},
                "pin": Object {
                  "django@1.6.1": Object {
                    "isTransitive": false,
                    "upgradeTo": "django@2.0.1",
                    "vulns": Array [
                      "vuln-id",
                    ],
                  },
                },
                "unresolved": Array [],
                "upgrade": Object {},
              },
            },
            "workspace": Object {
              "path": ".",
              "readFile": [Function],
              "writeFile": [Function],
            },
          },
        },
        Object {
          "changes": Array [
            Object {
              "from": "django@1.6.1",
              "issueIds": Array [
                "vuln-id",
              ],
              "success": true,
              "to": "django@2.0.1",
              "userMessage": "Upgraded django from 1.6.1 to 2.0.1",
            },
          ],
          "original": Object {
            "options": Object {
              "command": "python3",
            },
            "scanResult": Object {
              "facts": Array [
                Object {
                  "data": "not-implemented",
                  "type": "not-implemented",
                },
              ],
              "identity": Object {
                "targetFile": "requirements.txt",
                "type": "pip",
              },
            },
            "testResult": Object {
              "depGraphData": "",
              "issues": Array [
                Object {
                  "fixInfo": Object {},
                  "issueId": "VULN_ID_1",
                  "pkgName": "package@version",
                },
              ],
              "issuesData": Object {
                "vuln-id": Object {
                  "id": "VULN_ID_1",
                  "severity": "high",
                  "title": "Fake vuln",
                },
              },
              "remediation": Object {
                "ignore": Object {},
                "patch": Object {},
                "pin": Object {
                  "django@1.6.1": Object {
                    "isTransitive": false,
                    "upgradeTo": "django@2.0.1",
                    "vulns": Array [
                      "vuln-id",
                    ],
                  },
                },
                "unresolved": Array [],
                "upgrade": Object {},
              },
            },
            "workspace": Object {
              "path": ".",
              "readFile": [Function],
              "writeFile": [Function],
            },
          },
        },
      ],
    },
  },
}
`;

exports[`Snyk fix Snyk fix returns results for supported type 1`] = `Object {}`;

exports[`Snyk fix Snyk fix returns results for supported type 2`] = `
Object {
  "python": Object {
    "failed": Array [],
    "skipped": Array [],
    "succeeded": Array [
      Object {
        "changes": Array [
          Object {
            "from": "django@1.6.1",
            "issueIds": Array [
              "vuln-id",
            ],
            "success": true,
            "to": "django@2.0.1",
            "userMessage": "Upgraded django from 1.6.1 to 2.0.1",
          },
        ],
        "original": Object {
          "options": Object {
            "command": "python3",
          },
          "scanResult": Object {
            "facts": Array [
              Object {
                "data": "not-implemented",
                "type": "not-implemented",
              },
            ],
            "identity": Object {
              "targetFile": "requirements.txt",
              "type": "pip",
            },
          },
          "testResult": Object {
            "depGraphData": "",
            "issues": Array [
              Object {
                "fixInfo": Object {},
                "issueId": "VULN_ID_1",
                "pkgName": "package@version",
              },
            ],
            "issuesData": Object {
              "vuln-id": Object {
                "id": "VULN_ID_1",
                "severity": "high",
                "title": "Fake vuln",
              },
            },
            "remediation": Object {
              "ignore": Object {},
              "patch": Object {},
              "pin": Object {
                "django@1.6.1": Object {
                  "isTransitive": false,
                  "upgradeTo": "django@2.0.1",
                  "vulns": Array [
                    "vuln-id",
                  ],
                },
              },
              "unresolved": Array [],
              "upgrade": Object {},
            },
          },
          "workspace": Object {
            "path": ".",
            "readFile": [Function],
            "writeFile": [MockFunction] {
              "calls": Array [
                Array [
                  "requirements.txt",
                  "django===2.0.1",
                ],
              ],
              "results": Array [
                Object {
                  "type": "return",
                  "value": Promise {},
                },
              ],
            },
          },
        },
      },
    ],
  },
}
`;
